### PR TITLE
support emoji encoding for Flux jobids

### PR DESCRIPTION
**Pull request by grondo**
This PR implements the emoji encoding for FLUIDs as proposed in the pending PR flux-framework/rfc#381.
It is a WIP pending acceptance of that RFC PR.

Instead of trying to implement a generic binary to emoji encoding, the "basemoji" implementation here simply converts `uint64_t` to base 576 using the same pattern as the F58 implementation. This allows high-order zero bits to be dropped since the result is a "number" with emoji as the base 576 digits.

A named `emoji` encoding is then added to the `fluid.h` with support for detection and parsing in `fluid_parse(3)`, along with an `id.emoji` property for the `JobID` class and `flux-jobs` output formats.

```
$ flux job id --to=emoji 0 1 512 65535 123456789
😃
😄
🔞
💁📚
🎰💗📣
```

### CODE REVIEW

**Review by chu11 — COMMENTED** [review]
Idea: update the `cute` format to use `id.emoji`?  However, getting things to align is hard.  I haven't been able to figure anything out yet that looks ok.  Edit: I think the emojis are double width in my terminal, so I need to "space over" the JOBID header.  We need like a "header offset' or something.

image

File: src/common/libutil/basemoji.c [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+     * first two bytes match the expected UTF-8 encoding.
+     * This doesn't guarantee that `s` is a valid basemoji string,
+     * but this will catch most obvious cases and other invalid strings
+     * are left to be detected in decode.
+     */
+    if (len >= BASEMOJI_MINLEN
+        && len <= BASEMOJI_MAXLEN
+        && (uint8_t)s[0] == 0xf0
+        && (uint8_t)s[1] == 0x9f)
+        return true;
```
[the remark this answers is not in the document]
Comment (grondo) [reply src/common/libutil/basemoji.c:121]: Note we could go through and ensure ever other two bytes start with F0 9F, but that seemed a bit heavyweight to me. This isn't validating the emoji string, but just detecting if we should try to decode it as one..

File: src/common/libutil/test/fluid.c [hunk 2]
```diff
@@ -236,7 +245,7 @@ void test_basic (void)
     ok (fluid_encode (buf, sizeof (buf), id, FLUID_STRING_DOTHEX) == 0,
         "fluid_encode type=DOTHEX works");
     ok (fluid_decode (buf, &id2, FLUID_STRING_DOTHEX) == 0 && id == id2,
-        "fluid_decode type=MNEMONIC works");
+        "fluid_decode type=DOTHEX works");
```
[the remark this answers is not in the document]
Comment (grondo) [reply src/common/libutil/test/fluid.c:248]: Oh yeah, those were cut-and-paste fixes I missed adding to a different commit :-(

File: src/common/libutil/basemoji.c [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+     */
+    if ((count = emoji_revenc (reverse, sizeof (reverse), id)) < 0) {
+        errno = EINVAL;
+        return -1;
+    }
+
+    /*  Check for overflow of provided buffer:
+     *  Need space for prefix + count bytes for emoji + NUL
+     */
+    if (count + 1 > buflen) {
```
[the remark this answers is not in the document]
Comment (grondo) [reply src/common/libutil/basemoji.c:167]: Ok, but operator precedence puts `+` before `>` so what is the need here? Just preference?

[the remark this answers is not in the document]
Comment (garlick) [reply src/common/libutil/basemoji.c:167]: FWIW I would not add parens there.

**Comment by chu11** [comment]
I was completely messing around to see if I could get the emoji alignment to work and this ending up working:

```
@@ -443,6 +444,17 @@ class UtilFormatter(Formatter):
             basecases = ("", "0s", "0.0", "0:00:00", "1970-01-01T00:00:00", localepoch)
             value = "-" if str(value) in basecases else str(value)
             spec = spec[:-1] + "s"
+
+        if isinstance (value, str):
+            widecount = 0
+            for chr in value:
+                if unicodedata.east_asian_width(chr) == 'W':
+                    widecount += 1
+
+            if widecount and spec == ">12":
+                count = 12 - widecount
+                spec = f">{count}"
+
         retval = super().format_field(value, spec)
```

image

Obviously I hard coded it based on what I was expecting, but perhaps the basics of something to align wide chars is doable ... I think there's a lot of corner cases on emoji chars and some are half-chars, but the output would atleast be sensible in most cases.

**Comment by grondo** [comment]
Yeah, that might work! I wonder if we should have a special conversion so this doesn't have to be done for every string that is being formatted, or maybe there's some other way to better detect a possible "wide" string (or maybe it is fast enough that it doesn't matter)

**Comment by grondo** [comment]
BTW I stumbled across this library https://gitlab.com/fgallaire/cjkwrap/-/blob/master/cjkwrap.py
That has a `ckjlen()` function (len of a possible wide string) that we could maybe pilfer:
```python
def cjklen(text):
     """cjklen(object) -> integer
         Return the real width of an unicode text, the len of any other type.
     """
     if not isinstance(text, text_type):
         return len(text)
     return sum(2 if is_wide(char) else 1 for char in text) 
```

**Comment by grondo** [comment]
Just removed WIP since the RFC 19 changes went in.

**Comment by chu11** [comment]
> I wonder if we should have a special conversion so this doesn't have to be done for every string that is being formatted

Oh that's a good idea.  And honestly, we could just limit it to certain formatting in that case.  like `>\d+` and `<\d+` or something.

**Review by chu11 — APPROVED** [review]
LGTM

**Comment by chu11** [comment]
ehhh, this wasn't too awful, maybe could propose in a follow up PR (probably can cleanup some logic here).

```Python
diff --git a/src/bindings/python/flux/util.py b/src/bindings/python/flux/util.py
index f753f5e..1c2f82e 100644
--- a/src/bindings/python/flux/util.py
+++ b/src/bindings/python/flux/util.py
@@ -29,6 +29,7 @@ from string import Formatter
 from typing import Mapping
 
 import yaml
+import unicodedata
 
 # tomllib added to standard library in Python 3.11
 # flux-core minimum is Python 3.6.
@@ -443,6 +444,28 @@ class UtilFormatter(Formatter):
             basecases = ("", "0s", "0.0", "0:00:00", "1970-01-01T00:00:00", localepoch)
             value = "-" if str(value) in basecases else str(value)
             spec = spec[:-1] + "s"
+
+        if spec.endswith("W"):
+            if isinstance (value, str):
+                match = re.search(r"^([<>])(\d+)W", spec)
+                if match:
+                    align = match[1]
+                    width = int(match[2])
+                    widecount = 0
+                    for chr in value:
+                        if unicodedata.east_asian_width(chr) == 'W':
+                            widecount += 1
+
+                    if width > widecount:
+                        width -= widecount;
+                        spec = f"{align}{width}s"
+                    else:
+                        spec = ""
+                else:
+                    spec = spec[:-1]
+            else:
+                spec = spec[:-1]
+
         retval = super().format_field(value, spec)
 
         if denote_truncation and len(retval) < len(str(value)):
diff --git a/src/cmd/flux-jobs.py b/src/cmd/flux-jobs.py
index 5211fb6..a4a14bd 100755
--- a/src/cmd/flux-jobs.py
+++ b/src/cmd/flux-jobs.py
@@ -39,7 +39,7 @@ class FluxJobsConfig(UtilConfig):
         "cute": {
             "description": "Cute flux-jobs format string (default with emojis)",
             "format": (
-                "{id.f58:>12} ?:{queue:<8.8} {username:<8.8} {name:<10.10+} "
+                "{id.emoji:>12W} ?:{queue:<8.8} {username:<8.8} {name:<10.10+} "
                 "{status_emoji:>5.5} {ntasks:>6} {nnodes:>6h} "
                 "{contextual_time!F:>8h} {contextual_info}"
             ),
```

image

I did not add `W` to the `status_emoji` output though.  I think there is some combo of width, total width, and width of the header that is coming into play here for alignment that I haven't quite figured out.  I think this is a decent enough hack to get us something more sensible to update the `cute` format though.

**Comment by grondo** [comment]
Ok, I guess I'll set MWP then.. :laughing:

**Pull request merged**